### ISSUE TITLE
AppImage: The "magick++" output target is not available in Linux build

### ISSUE THREAD

**Issue by morevnaproject** [synfig/synfig#2320]
The "magick++" output target is not available in AppImage Linux builds for version 1.4.1, 1.4.2, 1.5.0.

**Comment by morevnaproject** [comment]
When launching Synfig Appimage from console it gives warning:
```
synfig(2628) [22:08:23] warning: Unable to find module "mod_magickpp" (file not found)
```

There is a `libmod_magickpp.so` exists in Appimage at `usr/lib/synfig/modules/`. 
Somehow Synfig doesn't finds it...

**Comment by ice0** [comment]
It cannot be loaded due to some dependencies (maybe it can't find `magick++.so` itself).
Unfortunately `lt_dlerror` is not showing the correct error :(

**Comment by morevnaproject** [comment]
The version compiled from source have magick++, so I have executed it with "strace" and got the following:

For compiled from source version:
```
$ strace /home/konstantin/sources/synfig/_production/build/bin/synfigstudio 2>&1 |grep magickpp
write(1, "synfig(11912) [22:21:12] info: L"..., 54synfig(11912) [22:21:12] info: Loading mod_magickpp..
write(1, "synfig(11912) [22:21:12] info: A"..., 69synfig(11912) [22:21:12] info: Attempting to register "mod_magickpp"
openat(AT_FDCWD, "/home/konstantin/.local/share/synfig/modules/libmod_magickpp.la", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/home/konstantin/sources/synfig/_production/build/lib/synfig/modules/libmod_magickpp.la", O_RDONLY) = 11
read(11, "# libmod_magickpp.la - a libtool"..., 4096) = 1403
openat(AT_FDCWD, "/home/konstantin/sources/synfig/_production/build/lib/synfig/modules/libmod_magickpp.so", O_RDONLY|O_CLOEXEC) = 11
write(1, "synfig(11912) [22:21:12] info: F"..., 59synfig(11912) [22:21:12] info: Found module "mod_magickpp"
write(1, "synfig(11912) [22:21:12] info: S"..., 58synfig(11912) [22:21:12] info: Success for "mod_magickpp"
```

For AppImage (I have extracted appimage and injected "strace" command to launcher script):
```
write(1, "synfig(12123) [22:24:32] info: L"..., 54synfig(12123) [22:24:32] info: Loading mod_magickpp..
write(1, "synfig(12123) [22:24:32] info: A"..., 69synfig(12123) [22:24:32] info: Attempting to register "mod_magickpp"
openat(AT_FDCWD, "/home/konstantin/.local/share/synfig/modules/libmod_magickpp.la", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/tmp/.mount_synfig/usr/lib/synfig/modules/libmod_magickpp.la", O_RDONLY) = 8
read(8, "# libmod_magickpp.la - a libtool"..., 4096) = 4096
openat(AT_FDCWD, "/build/packet/linux-64/synfigcore-master/install/lib/synfig/modules/libmod_magickpp.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/tmp/.mount_synfig/usr/lib/synfig/modules/libmod_magickpp.so", O_RDONLY|O_CLOEXEC) = 8
access("/home/konstantin/.local/share/synfig/modules/libmod_magickpp.so", R_OK) = -1 ENOENT (No such file or directory)
access("/tmp/.mount_synfig/usr/lib/synfig/modules/libmod_magickpp.so", R_OK) = 0
openat(AT_FDCWD, "/tmp/.mount_synfig/usr/lib/synfig/modules/libmod_magickpp.so", O_RDONLY|O_CLOEXEC) = 8
openat(AT_FDCWD, "/home/konstantin/.local/share/synfig/modules/mod_magickpp.la", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/tmp/.mount_synfig/usr/lib/synfig/modules/mod_magickpp.la", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/build/packet/linux-64/synfigcore-master/install/lib/synfig/modules/mod_magickpp.la", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "./mod_magickpp.la", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/tmp/.mount_synfig/usr/lib/mod_magickpp.la", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/tmp/.mount_synfig/usr/lib64/mod_magickpp.la", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/lib/mod_magickpp.la", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/lib/mod_magickpp.la", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/lib/x86_64-linux-gnu/libfakeroot/mod_magickpp.la", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/local/lib/mod_magickpp.la", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/lib/x86_64-linux-gnu/mod_magickpp.la", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/lib/x86_64-linux-gnu/mod_magickpp.la", O_RDONLY) = -1 ENOENT (No such file or directory)
access("/home/konstantin/.local/share/synfig/modules/mod_magickpp.so", R_OK) = -1 ENOENT (No such file or directory)
access("/tmp/.mount_synfig/usr/lib/synfig/modules/mod_magickpp.so", R_OK) = -1 ENOENT (No such file or directory)
access("/build/packet/linux-64/synfigcore-master/install/lib/synfig/modules/mod_magickpp.so", R_OK) = -1 ENOENT (No such file or directory)
access("./mod_magickpp.so", R_OK)       = -1 ENOENT (No such file or directory)
access("/tmp/.mount_synfig/usr/lib/mod_magickpp.so", R_OK) = -1 ENOENT (No such file or directory)
access("/tmp/.mount_synfig/usr/lib64/mod_magickpp.so", R_OK) = -1 ENOENT (No such file or directory)
access("/lib/mod_magickpp.so", R_OK)    = -1 ENOENT (No such file or directory)
access("/usr/lib/mod_magickpp.so", R_OK) = -1 ENOENT (No such file or directory)
access("/usr/lib/x86_64-linux-gnu/libfakeroot/mod_magickpp.so", R_OK) = -1 ENOENT (No such file or directory)
access("/usr/local/lib/mod_magickpp.so", R_OK) = -1 ENOENT (No such file or directory)
access("/lib/x86_64-linux-gnu/mod_magickpp.so", R_OK) = -1 ENOENT (No such file or directory)
access("/usr/lib/x86_64-linux-gnu/mod_magickpp.so", R_OK) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/tmp/.mount_synfig/usr/lib/mod_magickpp.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/tmp/.mount_synfig/usr/lib64/mod_magickpp.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "tls/x86_64/x86_64/mod_magickpp.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "tls/x86_64/mod_magickpp.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "tls/x86_64/mod_magickpp.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "tls/mod_magickpp.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "x86_64/x86_64/mod_magickpp.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "x86_64/mod_magickpp.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "x86_64/mod_magickpp.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "mod_magickpp.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/lib/x86_64-linux-gnu/mod_magickpp.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/lib/x86_64-linux-gnu/tls/mod_magickpp.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/lib/x86_64-linux-gnu/mod_magickpp.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/lib/mod_magickpp.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/usr/lib/mod_magickpp.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
write(2, "Unable to find module \"mod_magic"..., 53Unable to find module "mod_magickpp" (file not found)) = 53
```

Notice the lines in first case (file found and working):
```
openat(AT_FDCWD, "/home/konstantin/sources/synfig/_production/build/lib/synfig/modules/libmod_magickpp.la", O_RDONLY) = 11
read(11, "# libmod_magickpp.la - a libtool"..., 4096) = 1403
openat(AT_FDCWD, "/home/konstantin/sources/synfig/_production/build/lib/synfig/modules/libmod_magickpp.so", O_RDONLY|O_CLOEXEC) = 11
```

In second case (file exists, but not loaded):
```
openat(AT_FDCWD, "/tmp/.mount_synfig/usr/lib/synfig/modules/libmod_magickpp.la", O_RDONLY) = 8
read(8, "# libmod_magickpp.la - a libtool"..., 4096) = 4096
openat(AT_FDCWD, "/build/packet/linux-64/synfigcore-master/install/lib/synfig/modules/libmod_magickpp.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/tmp/.mount_synfig/usr/lib/synfig/modules/libmod_magickpp.so", O_RDONLY|O_CLOEXEC) = 8
```

**Comment by morevnaproject** [comment]
Appimage version 1.4.0 loads magick++ successfully. Here is strace log:

```
konstantin@master:~/sources/synfig$ '/tmp/.mount_synfig140/usr/bin/launch.sh' 
write(1, "synfig(13227) [22:49:07] info: L"..., 54synfig(13227) [22:49:07] info: Loading mod_magickpp..
write(1, "synfig(13227) [22:49:07] info: A"..., 69synfig(13227) [22:49:07] info: Attempting to register "mod_magickpp"
openat(AT_FDCWD, "/home/konstantin/.local/share/synfig/modules/libmod_magickpp.la", O_RDONLY) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/tmp/.mount_synfig140/usr/lib/synfig/modules/libmod_magickpp.la", O_RDONLY) = 8
read(8, "# libmod_magickpp.la - a libtool"..., 4096) = 4096
openat(AT_FDCWD, "/build/packet/linux-64/synfigcore-master/install/lib/synfig/modules/libmod_magickpp.so", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/tmp/.mount_synfig140/usr/lib/synfig/modules/libmod_magickpp.so", O_RDONLY|O_CLOEXEC) = 8
write(1, "synfig(13227) [22:49:07] info: F"..., 59synfig(13227) [22:49:07] info: Found module "mod_magickpp"
write(1, "synfig(13227) [22:49:07] info: S"..., 58synfig(13227) [22:49:07] info: Success for "mod_magickpp"
```

**Comment by morevnaproject** [comment]
Okay, I took two appimages - 1.4.0 (with working magick++) and 1.4.1 (with non-working magick++).
I unpacked both appimages.

Then I copied "libmod_magickpp.so" module from 1.4.0 appimage dir to 1.4.1 appimage dir.
Then I launched 1.4.1 version - the magick++ module loaded.

I guess this tells that it's not the problem of AppImage environment, and not problem of Synfig code.

So, I have examined diff between code of 1.4.0 vs code of 1.4.1 and the only suspicious thing I see is this O_O -
![screenshot_003](https://user-images.githubusercontent.com/332868/134039027-9b49c6c5-7d0a-45f0-8d6e-372d85db6d20.png)

**Comment by morevnaproject** [comment]
I have made a test commit and will try to rebuild AppImage to see if it makes any difference.

**Comment by ice0** [comment]
@morevnaproject 

Just go to folder where `libmod_magickpp.so` is located and use `ldd` with correct `LD_LIBRARY_PATH`:

```
LD_LIBRARY_PATH=/home/user/synfig-fix-appimage/1.5.0/usr/lib ldd ./libmod_magickpp.so
	linux-vdso.so.1 (0x00007ffceb5e7000)
	libMagick++-6.Q16.so.3 => /home/user/synfig-fix-appimage/1.5.0/usr/lib/libMagick++-6.Q16.so.3 (0x00007f0f4d809000)
	libMagickWand-6.Q16.so.1 => /home/user/synfig-fix-appimage/1.5.0/usr/lib/libMagickWand-6.Q16.so.1 (0x00007f0f4d467000)
	libMagickCore-6.Q16.so.1 => /home/user/synfig-fix-appimage/1.5.0/usr/lib/libMagickCore-6.Q16.so.1 (0x00007f0f4d033000)
	libfreetype.so.6 => /usr/lib/x86_64-linux-gnu/libfreetype.so.6 (0x00007f0f4cf57000)
	libfontconfig.so.1 => /usr/lib/x86_64-linux-gnu/libfontconfig.so.1 (0x00007f0f4cf0e000)
	libbz2.so.1.0 => /home/user/synfig-fix-appimage/1.5.0/usr/lib/libbz2.so.1.0 (0x00007f0f4ccfe000)
	libgomp.so.1 => /home/user/synfig-fix-appimage/1.5.0/usr/lib/libgomp.so.1 (0x00007f0f4cace000)
	libsynfig.so.0 => /home/user/synfig-fix-appimage/1.5.0/usr/lib/libsynfig.so.0 (0x00007f0f4c1c6000)
	libltdl.so.7 => /usr/lib/x86_64-linux-gnu/libltdl.so.7 (0x00007f0f4c1bb000)
	libgiomm-2.4.so.1 => /home/user/synfig-fix-appimage/1.5.0/usr/lib/libgiomm-2.4.so.1 (0x00007f0f4bd36000)
	libgio-2.0.so.0 => /home/user/synfig-fix-appimage/1.5.0/usr/lib/libgio-2.0.so.0 (0x00007f0f4b945000)
	libselinux.so.1 => /home/user/synfig-fix-appimage/1.5.0/usr/lib/libselinux.so.1 (0x00007f0f4b725000)
	libresolv.so.2 => /lib/x86_64-linux-gnu/libresolv.so.2 (0x00007f0f4b709000)
	libmount.so.1 => /lib/x86_64-linux-gnu/libmount.so.1 (0x00007f0f4b6a9000)
	libxml++-2.6.so.2 => /home/user/synfig-fix-appimage/1.5.0/usr/lib/libxml++-2.6.so.2 (0x00007f0f4b44b000)
	libxml2.so.2 => /home/user/synfig-fix-appimage/1.5.0/usr/lib/libxml2.so.2 (0x00007f0f4b078000)
	libz.so.1 => /home/user/synfig-fix-appimage/1.5.0/usr/lib/libz.so.1 (0x00007f0f4ae59000)
	liblzma.so.5 => /lib/x86_64-linux-gnu/liblzma.so.5 (0x00007f0f4ae30000)
	libglibmm-2.4.so.1 => /home/user/synfig-fix-appimage/1.5.0/usr/lib/libglibmm-2.4.so.1 (0x00007f0f4ab29000)
	libgmodule-2.0.so.0 => /home/user/synfig-fix-appimage/1.5.0/usr/lib/libgmodule-2.0.so.0 (0x00007f0f4a926000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f0f4a920000)
	libgobject-2.0.so.0 => /home/user/synfig-fix-appimage/1.5.0/usr/lib/libgobject-2.0.so.0 (0x00007f0f4a6c8000)
	libffi.so.5 => not found
	libglib-2.0.so.0 => /home/user/synfig-fix-appimage/1.5.0/usr/lib/libglib-2.0.so.0 (0x00007f0f4a357000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f0f4a334000)
	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f0f4a329000)
	libfftw3.so.3 => /home/user/synfig-fix-appimage/1.5.0/usr/lib/libfftw3.so.3 (0x00007f0f49f41000)
	libmlt++.so.3 => /home/user/synfig-fix-appimage/1.5.0/usr/lib/libmlt++.so.3 (0x00007f0f49d1d000)
	libmlt.so.6 => /home/user/synfig-fix-appimage/1.5.0/usr/lib/libmlt.so.6 (0x00007f0f49ae4000)
	libsigc-2.0.so.0 => /home/user/synfig-fix-appimage/1.5.0/usr/lib/libsigc-2.0.so.0 (0x00007f0f498d1000)
	libstdc++.so.6 => /home/user/synfig-fix-appimage/1.5.0/usr/lib/libstdc++.so.6 (0x00007f0f4953a000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f0f493eb000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f0f491f9000)
	libgcc_s.so.1 => /home/user/synfig-fix-appimage/1.5.0/usr/lib/libgcc_s.so.1 (0x00007f0f48fe2000)
	libpng16.so.16 => /home/user/synfig-fix-appimage/1.5.0/usr/lib/libpng16.so.16 (0x00007f0f48d9e000)
	libexpat.so.1 => /lib/x86_64-linux-gnu/libexpat.so.1 (0x00007f0f48d70000)
	libuuid.so.1 => /lib/x86_64-linux-gnu/libuuid.so.1 (0x00007f0f48d67000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f0f4dcca000)
	libffi.so.6 => not found
	libffi.so.6 => not found
	libblkid.so.1 => /lib/x86_64-linux-gnu/libblkid.so.1 (0x00007f0f48d0e000)
	libffi.so.6 => not found
```

as you can notice:
```
	libffi.so.5 => not found
	libffi.so.6 => not found
```

1) Is it linked with two different `libffi` versions?
2) `libffi.so.6` is located under `lib64` folder, not in `lib` folder. So you need to fix that output path.

**Comment by ice0** [comment]
> Is it linked with two different `libffi` versions?

It looks like you have two different `libffi` libraries on your build system. Before we added `@MAGICKPP_CFLAGS@`, it just took one of them. So you need to check `PKG_CONFIG_PATH` and link flags in your build environment.

**Comment by morevnaproject** [comment]
You right, in  AppImage for version 1.4.0 it doesn't references `libffi.so.5`, but in AppImage for 1.4.1 it references both `libffi.so.5` and `libffi.so.6`.
On my system `libffi.so.6` is found at 
```
libffi.so.6 => /usr/lib/x86_64-linux-gnu/libffi.so.6
```
Still, Synfig's launcher script sets LD_LIBRARY_PATH to look into `lib64/` dir too (inside of AppImage). SO I think this shouldn't be the problem.

I still think the library linked differently because the change made to `mod_magickpp/Makefile.am`.
It is done in this commit - a583e73d4c47914284b15f4a8d7719eafddfe1b4

Notice it moves `$MAGICKPP_CFLAGS` from `synfig-core/configure.ac` to `mod_magickpp/Makefile.am`.
But `$MAGICKPP_LIBS` variable is not moved to `mod_magickpp/Makefile.am` (I believe it should be present in `libmod_magickpp_la_LIBADD` list).
So, I have made the test change in https://github.com/morevnaproject/synfig/commit/27b3500e3aff9e82e71ed9361fd49f352f72bda3 and rebuilding appimage to see if it will work. ^__^

**Comment by morevnaproject** [comment]
> So, I have made the test change in [morevnaproject@4252558](https://github.com/morevnaproject/synfig/commit/4252558bb11bc488950feba7f5468ddfc157ae21) and rebuilding appimage to see if it will work. ^__^

It works now! - https://coolbug.org/downloads/morevna/builds/SynfigStudio-1.7.0-testing-2021.09.21-linux64-27b35.appimage

**Comment by morevnaproject** [comment]
Resolved in Synfig version 1.5.1 - https://www.synfig.org/2021/10/30/synfig-studio-1-5-1-released/
